ts/session/sending/MessageSender.ts

```diff
@@ -2,13 +2,110 @@
 
 import { RawMessage } from '../types/RawMessage';
 import { OpenGroupMessage } from '../messages/outgoing';
+import { SignalService } from '../../protobuf';
+import { UserUtil } from '../../util';
+import { MessageEncrypter } from '../crypto';
+import { lokiMessageAPI, lokiPublicChatAPI, textsecure } from '../../window';
 
-export async function send(message: RawMessage): Promise<void> {
-  return Promise.resolve();
+// ================ Regular ================
+
+export function canSendToSnode(): boolean {
+  // Seems like lokiMessageAPI is not always guaranteed to be initialized
+  return Boolean(lokiMessageAPI);
+}
+
+export async function send({
+  device,
+  plainTextBuffer,
+  encryption,
+  timestamp,
+  ttl,
+}: RawMessage): Promise<void> {
+  if (!canSendToSnode()) {
+    throw new Error('lokiMessageAPI is not initialized.');
+  }
+
+  const { envelopeType, cipherText } = await MessageEncrypter.encrypt(
+    device,
+    plainTextBuffer,
+    encryption
+  );
+  const envelope = await buildEnvelope(envelopeType, timestamp, cipherText);
+  const data = wrapEnvelope(envelope);
+
+  // TODO: Somehow differentiate between Retryable and Regular erros
+  return lokiMessageAPI.sendMessage(device, data, timestamp, ttl);
+}
+
+async function buildEnvelope(
+  type: SignalService.Envelope.Type,
+  timestamp: number,
+  content: Uint8Array
+): Promise<SignalService.Envelope> {
+  let source: string | undefined;
+  if (type !== SignalService.Envelope.Type.UNIDENTIFIED_SENDER) {
+    source = await UserUtil.getCurrentDevicePubKey();
+  }
+
+  return SignalService.Envelope.create({
+    type,
+    source,
+    sourceDevice: 1,
+    timestamp,
+    content,
+  });
 }
 
+/**
+ * This is an outdated practice and we should probably just send the envelope data directly.
+ * Something to think about in the future.
+ */
+function wrapEnvelope(envelope: SignalService.Envelope): Uint8Array {
+  const request = SignalService.WebSocketRequestMessage.create({
+    id: 0,
```

Comment:
existing code was this:
`    id: new Uint8Array(libsignal.crypto.getRandomBytes(1))[0], // random ID for now`

We do not care anymore?

Reply:
Well we don't really use the id for much so I figured hardcoding it wouldn't matter.